velox/functions/prestosql/json/JsonStringUtil.cpp

```diff
@@ -0,0 +1,172 @@
+
+#include "velox/common/base/Exceptions.h"
+#include "velox/functions/lib/Utf8Utils.h"
+#include "velox/functions/prestosql/json/JsonStringUtil.h"
+
+using namespace facebook::velox::functions;
+
+namespace facebook::velox {
+namespace {
+
+FOLLY_ALWAYS_INLINE char hexDigit(uint8_t c) {
+  VELOX_DCHECK_LT(c, 16);
+  return c < 10 ? c + '0' : c - 10 + 'A';
+}
+
+FOLLY_ALWAYS_INLINE void writeHex(char16_t value, char*& out) {
+  value = folly::Endian::little(value);
+  *out++ = '\\';
+  *out++ = 'u';
+  *out++ = hexDigit((value >> 12) & 0x0F);
+  *out++ = hexDigit((value >> 8) & 0x0F);
+  *out++ = hexDigit((value >> 4) & 0x0F);
+  *out++ = hexDigit(value & 0x0F);
+}
+
+std::array<int8_t, 128> getAsciiEscapes() {
+  std::array<int8_t, 128> escapes;
+  std::fill(escapes.data(), escapes.data() + 32, -1);
+  escapes['"'] = '"';
+  escapes['\\'] = '\\';
+  escapes['\b'] = 'b';
+  escapes['\t'] = 't';
+  escapes['\n'] = 'n';
+  escapes['\f'] = 'f';
+  escapes['\r'] = 'r';
+  return escapes;
+}
+static const std::array<int8_t, 128> asciiEscapes = getAsciiEscapes();
+
+FOLLY_ALWAYS_INLINE void encodeAscii(int8_t value, char*& out) {
+  int8_t escapeCode = asciiEscapes[value];
+  if (escapeCode == 0) {
+    *out++ = char(value);
+  } else if (escapeCode > 0) {
+    *out++ = '\\';
+    *out++ = char(escapeCode);
+  } else {
+    writeHex(value, out);
+  }
+}
+
+FOLLY_ALWAYS_INLINE size_t encodedAsciiSize(uint8_t value) {
+  int8_t escapeCode = asciiEscapes[value];
+  if (escapeCode == 0) {
+    return 1;
+  }
+  if (escapeCode > 0) {
+    return 2;
+  }
+  return 6;
+}
+
+// Encode `codePoint` value into one or two UTF-16 code units. Write each code
+// unit as prefixed hexadecimals of 6 chars.
+FOLLY_ALWAYS_INLINE void encodeUtf16Hex(char32_t codePoint, char*& out) {
+  VELOX_DCHECK(codePoint <= 0x10FFFFu);
+  // Two 16-bit code units are needed.
+  if (codePoint >= 0x10000u) {
+    writeHex(
+        static_cast<char16_t>(
+            0xD800u + (((codePoint - 0x10000u) >> 10) & 0x3FFu)),
+        out);
+    writeHex(
+        static_cast<char16_t>(0xDC00u + ((codePoint - 0x10000u) & 0x3FFu)),
+        out);
+    return;
+  }
+  // One 16-bit code unit is needed.
+  writeHex(static_cast<char16_t>(codePoint), out);
+}
+
+} // namespace
+
+void testingEncodeUtf16Hex(char32_t codePoint, char*& out) {
+  encodeUtf16Hex(codePoint, out);
+}
+
+void escapeString(const char* input, size_t length, char* output) {
+  char* pos = output;
+
+  *pos++ = '"';
+
+  auto* start = reinterpret_cast<const unsigned char*>(input);
+  auto* end = reinterpret_cast<const unsigned char*>(input + length);
+  while (start < end) {
+    int count = validateAndGetNextUtf8Length(start, end);
+    if (count == -1) {
+      writeHex(0xFFFDu, pos);
+      start++;
+      continue;
+    }
+    if (count == 1) {
+      encodeAscii(int8_t(*start), pos);
+      start++;
+      continue;
+    }
+    if (count == 2 || count == 3) {
+      for (int i = 0; i < count; i++) {
```

Comment:
In these cases just write it out without a loop would be more efficient:
```c++
switch (count) {
case 2:
  memcpy(pos, start, 2);
  pos += 2;
  start += 2;
  break;
case 3:
  memcpy(pos, start, 3);
  position += 3;
  start += 3;
  break;
}
```

Reply:
Changed. Thought about memcpy before.